### ISSUE TITLE
enhancement: `dict[key] ?? default` as syntactic sugar for `get(()->default, dict, key) ` (with analogous uses for `nullableValue ?? default` and for `?=`)

### ISSUE THREAD

**Issue by jamesonquinn** [JuliaLang/julia#23308]
This is an idea that came up in the discussion of https://github.com/JuliaLang/julia/issues/23275#issuecomment-322990685 and that probably merits its own separate discussion.

This same syntax would also work for other containers such as Arrays (`myarray[i,j,k; default]`, Nullables (`mynullable[; default]`), etc. If you wanted to make it a generalized language feature, you could potentially think about crazy additions like `mydict[key; default=0., use_hash_only=true]` but I think that it's probably better to KISS.

**Comment by StefanKarpinski** [comment]
I agree that we need some syntax for this, but I don't care for this one. The reason I think we need syntax is that there are often cases where you don't want to evaluate the default expression, which is currently rather awkward to express – `get(()->default, dict, key)` for lookup with default that is only evaluated on demand, and `get!(()->default, dict, key)` for the assigning version of that.

Some syntax that I've considered:
```jl
dict[key] ?? default
# equivalent to
get(()->default, dict, key)

dict[key] ?= default
# equivalent to
get!(()->default, dict, key)
```
and for non-indexing expressions, perhaps:
```jl
value ?? default
# equivalent to
(@isdefined(value) && value != nothing) ? value : default

value ?= default
# equivalent to
(@isdefined(value) && value != nothing) ? value : (value = default)
```
Note that the dict version is syntactically similar to the scalar version, but doesn't involve `nothing` as a sentinel value.

**Comment by jamesonquinn** [comment]
I think Stefan's proposal is better than my original one.

Stefan's does have an interaction with the Null issue. It seems as if the plan is to have the datalike null be `Union{T,Void}`, which would work with this; but for the CSish null, there are various proposals. Of those, `Union{Some{T},Void}` would work with this, but that proposal brings up issues with confusing semantics of raising; we want `2+missableIntWhichHappensToBeNa` to be NA (`na`?), but `2+nullableIntWhichHappensToBeNull` to raise a missing method error.

So, I'd say that `value ?? default` should be `(@isdefined(value) && hasvalue(value) ? value : default`, where `hasvalue(x)` is a new base method which is true for `x::Any`, false for `x::Void`, and `!isnull(x)` for `x::Nullable`. (Obviously, if needed, replace `Nullable` with `Maybe` or whatever we decide the final name is.)

**Comment by jamesonquinn** [comment]
One more question worth thinking about: what about `[2,3][4] ?? 5`? Intuitively, that's similar to `dict[key] ?? default`, but if it were going to actually mean anything, it would have to be by way of an implicit try/catch block; and I don't see any way for the corresponding `?=` to mean anything at all.

This could be done using something like my `hasvalue` idea above, except it would have to be changed to `@hasvalue`, which would make extending it decidedly harder and more dangerous.

The other way for this to have a meaning would be to define `get(::Function,::AbstractArray,::Any)` with the appropriate try/catch block. 

For now, I think this may be a bridge too far. But it's worth considering to see if there's a good solution.

**Comment by StefanKarpinski** [comment]
I'm against an extensible `hasvalue` since you then get into weird "truthiness" business where everyone's playing some guessing game about what someone else decided was a value or not. If we have a standard not-a-value sentinel, then it should be standard. Programmability of `haskey` is a different situation since that function's purpose is literally to check if a collection has an index.

**Comment by jamesonquinn** [comment]
Changing title from "enhancement: `mydict[key; default]` as syntactic sugar for `get(mydict, key, default)`." to "enhancement: `dict[key] ?? default` as syntactic sugar for `get(()->default, dict, key)
` (with analogous uses for `nullableValue ?? default` and for `?=`)"

**Comment by cstjean** [comment]
> Some syntax that I've considered:

If there was a nicer syntax for one-liner do notation, `dict[key] do; default; end` would work.

**Comment by StefanKarpinski** [comment]
Do syntax is already a nicer syntax for passing an anonymous function, so I'm not sure about that.

**Comment by StefanKarpinski** [comment]
I would still love this but I think in the minority.

**Comment by adienes** [comment]
oops, thought the close would be noncontroversial. I did not mean to be making any executive decisions

I should also note that, sadly, the OP of this issue has passed away.

**Comment by StefanKarpinski** [comment]
Oh, that is sad 😔
